File: src/main/java/processes/ProcessHandleFromString.java
package processes;

import java.time.Duration;
import java.time.Instant;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Stream;

class ProcessHandleFromString implements ProcessHandle {
    private final String command;
    private final String user;

    public ProcessHandleFromString(String command, String user) {
        this.command = command;
        this.user = user;
    }

    public ProcessHandleFromString(String command) {
        this(command,"apoc");
    }


    @Override
    public long pid() {
        return 0;
    }

    @Override
    public Optional<ProcessHandle> parent() {
        return Optional.empty();
    }

    @Override
    public Stream<ProcessHandle> children() {
        return null;
    }

    @Override
    public Stream<ProcessHandle> descendants() {
        return null;
    }

    @Override
    public ProcessHandle.Info info() {
        return new ProcessHandle.Info() {
            @Override
            public Optional<String> command() {
                return Optional.of(command);
            }

            @Override
            public Optional<String> commandLine() {
                return Optional.empty();
            }

            @Override
            public Optional<String[]> arguments() {
                return Optional.empty();
            }

            @Override
            public Optional<Instant> startInstant() {
                return Optional.empty();
            }

            @Override
            public Optional<Duration> totalCpuDuration() {
                return Optional.empty();
            }

            @Override
            public Optional<String> user() {
                return Optional.of(user);
            }
        };
    }

    @Override
    public CompletableFuture<ProcessHandle> onExit() {
        return null;
    }

    @Override
    public boolean supportsNormalTermination() {
        return false;
    }

    @Override
    public boolean destroy() {
        return false;
    }

    @Override
    public boolean destroyForcibly() {
        return false;
    }

    @Override
    public boolean isAlive() {
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(info().command(), info().user());
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof ProcessHandle)) return false;
        ProcessHandle process = (ProcessHandle) other;
        return info().command().orElse("").equals(process.info().command().orElse("")) &&
                info().user().orElse("").equals(process.info().user().orElse(""));
    }

    @Override
    public int compareTo(ProcessHandle other) {
        int dontLetItBeEqual = this.info().command().orElse("").compareTo(other.info().command().orElse(""))
                + this.info().user().orElse("").compareTo(other.info().user().orElse(""));
        if (dontLetItBeEqual == 0 && !this.equals(other)) return -1;
        return dontLetItBeEqual;
    }
}

File: src/main/kotlin/processes/ProcessHandleFromString.kt
package processes

import java.time.Duration
import java.time.Instant
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.stream.Stream

internal class ProcessHandleFromString @JvmOverloads constructor(
    private val command: String,
    private val user: String = "apoc"
) :
    ProcessHandle {
    override fun pid(): Long {
        return 0
    }

    override fun parent(): Optional<ProcessHandle> {
        return Optional.empty()
    }

    override fun children(): Stream<ProcessHandle> {
        return Stream.empty()
    }

    override fun descendants(): Stream<ProcessHandle> {
        return Stream.empty()
    }

    override fun info(): ProcessHandle.Info {
        return object : ProcessHandle.Info {
            override fun command(): Optional<String> {
                return Optional.of(command)
            }

            override fun commandLine(): Optional<String> {
                return Optional.empty()
            }

            override fun arguments(): Optional<Array<String>> {
                return Optional.empty()
            }

            override fun startInstant(): Optional<Instant> {
                return Optional.empty()
            }

            override fun totalCpuDuration(): Optional<Duration> {
                return Optional.empty()
            }

            override fun user(): Optional<String> {
                return Optional.of(user)
            }
        }
    }

    override fun onExit(): CompletableFuture<ProcessHandle> {
        return CompletableFuture()
    }

    override fun supportsNormalTermination(): Boolean {
        return false
    }

    override fun destroy(): Boolean {
        return false
    }

    override fun destroyForcibly(): Boolean {
        return false
    }

    override fun isAlive(): Boolean {
        return true
    }

    override fun hashCode(): Int {
        return Objects.hash(info().command(), info().user())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ProcessHandle) return false
        val process = other
        return info().command().orElse("") == process.info().command().orElse("") && info().user()
            .orElse("") == process.info().user().orElse("")
    }

    override fun compareTo(other: ProcessHandle): Int {
        val dontLetItBeEqual = (info().command().orElse("").compareTo(other.info().command().orElse(""))
                + info().user().orElse("").compareTo(other.info().user().orElse("")))
        return if (dontLetItBeEqual == 0 && this != other) -1 else dontLetItBeEqual
    }
}